docs/source/guides/testing-react-apollo.md

```diff
@@ -0,0 +1,334 @@
+      },
+    },
+    result: {
+      data: {
+        dog: { id: '1', name: 'Buck', breed: 'bulldog' },
+      },
+    },
+  },
+];
+
+it('renders without error', () => {
+  renderer.create(
+    <MockedProvider mocks={mocks} addTypename={false}>
+      <Dog name="Buck" />
+    </MockedProvider>,
+  );
+});
+```
+
+This example shows how to define the mocked response for a query. The `mocks` array takes objects with specific `request`s and their associated `result`s. In this example, when the provider receives a `GET_DOG_QUERY` with the specified variables, it returns the object under the `result` key.
+
+### addTypename
+
+You may notice the prop being passed to the `MockedProvider` called `addTypename`. The reason this is here is because of how Apollo Client normally works. When a request is made with Apollo Client normally, it adds a `__typename` field to every object type requested. This is to make sure that Apollo Client's cache knows how to normalize and store the response. When we're making our mocks, though, we're importing the raw queries _without typenames_ from the component files.
+
+If we don't disable the adding of typenames to queries, our imported query won't match the query actually being run by the component during our tests.
+
+> In short, If you don't have `__typename`s in your queries, pass `addTypename={false}` to your `MockedResolver`s.
+
+## Testing loading states
+
+In this example, the `Dog` component will render, but it will render in a loading state, not the final response state. This is because `MockedProvider` doesn't just return the data. It returns a promise that will resolve to that data. This allows testing of loading states as well as the final state.
+
+```js
+it('should render loading state initially', () => {
+  const component = renderer.create(
+    <MockedProvider mocks={[]}>
+      <Dog />
+    </MockedProvider>,
+  );
+
+  const tree = component.toJSON();
+  expect(tree.children).toContain('Loading...');
+});
+```
+
+This example shows how to (very basically) test loading state of a component. Here, all we're checking is that the children of the component contain the text `Loading...`. In real apps, this test would probably be more complicated, but the testing logic would be the same.
+
+## Testing final state
+
+Loading state, while important, isn't the only thing to test. To test the final state of the component after receiving data, we can just wait for it to update and test the final state.
+
+```js
+const wait = require('waait');
+
+it('should render currency conversions', async () => {
+  const dogMock = {
+    request: {
+      query: GET_DOG_QUERY,
+      variables: { name: 'Buck' },
+    },
+    result: {
+      data: { dog: { id: 1, name: 'Buck', breed: 'poodle' } },
+    },
+  };
+
+  const component = renderer.create(
+    <MockedProvider mocks={[dogMock]} addTypename={false}>
+      <Dog name="Buck" />
+    </MockedProvider>,
+  );
+
+  await wait(0); // wait for response
```

Comment:
Does this always work consistently? As I remember correctly adding a timeout of 0 doesn't always guarantee that the loading state is gone. Therefore in the `react-apollo` [examples](https://github.com/apollographql/react-apollo/pull/1994/files) I used [wait-for-expect](https://github.com/TheBrainFamily/wait-for-expect) to test the components and this proved to work great:) What do you think?

```js
import wait from "wait-for-expect";

it("tests", () => {
// This statement will wait until loading UI disappears before continuing
await wait(() => {
    isNotLoading();
})

// Data is present 

});
```

Reply:
Oh, interesting. I'll have to check out those examples and see. I never saw the loading state not be gone (although these examples are very simple). 

Otherwise, I like wait-for-expect. I'll check it out and see if it's a simple enough solution!

Reply:
Great, thanks a lot for taking it into consideration:)